docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+spec:
+  endpoint:
+    host:
+    port:
+status:
+  targetKind: <string representing group/kind name for selector>
+  targetSelector: <string representing the label selector for the referenced LB>
+  targets: (List of members matching label selector)
+  activeTargets: (List of active members)
+```
+
+#### Conditions
+
+- DNSResolvable
+- LBExists
+- LBMatchesConfig
+- TODO: condition to indicate ready to unblock things like ControlPlane initialization
+- TODO: condition to indicate fully healthy or degraded
+
+#### Watches:
+
+- own resource
+- Kind referenced by LB selector (to identify changes to "disabled" annotation)
+- LoadBalancer resources associated with own resource
+
+#### Reconciliation
+
+![FooLoadBalancer Reconciliation Flow](images/load-balancer/foolb-reconciliation.png)
+
+
+- If Spec.Endpoint is not defined:
+    - Create new LB using provider defaults for endpoint, populate Spec.Endpoint
+- Lookup related LoadBalancer resource:
+    - For any matching resources:
+        - resource is not deleting:
+            - `loadbalancer.cluster.x-k8s.io/disabled` not present:
+                - resource not present in load balancer:
+                    - add resource to load balancer config
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'adding'
+                    - requeue resource appropriately
+                - resource present in load balancer, but not active:
+                    - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                    - requeue resource appropriately
+                - resource present and active in load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'active'
+            - `loadbalancer.cluster.x-k8s.io/disabled` is present:
+                - resource active in load balancer:
+                    - start removal/disabling of resource from lb
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabling'
+                    - requeue resource appropriately
+                - resource is being removed/disabled from lb:
+                    - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                    - requeue resource appropriately
+                - resource is removed/disabled from the load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabled'
+        - resource is deleting:
+            - resource active in load balancer:
+                - start removal of resource from lb:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removing'
+                - requeue resource appropriately
+            - resource is being removed from lb:
+                - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                - requeue resource appropriately
+            - resource is removed from load balancer:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removed'
```

Comment:
In general, I don't think it matters much if it was removed or never a member of from the perspective of the controller implementations, but I think it does provide value from the perspective of a user trying to troubleshoot why things aren't working as expected.

There are a few different purposes for the annotation, though.
1) Allow conveying individual member status from the load balancer infrastructure provider to the general load balancer controller for the purpose of common reconciliation logic
2) Allow for users/controllers to filter resources based on various provider states for the purposes of tracking operations or troubleshooting issues.
3) Since we cannot know all potential intermediate steps for underlying implementations, this annotation gives us a way to generalize the common states enough for gating purposes and still allows the providers to expose state that can be summarized in useful ways for users.